compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+      rewriter.create<linalg::YieldOp>(loc, truncOp);
+    };
+    auto truncateOp = rewriter.create<linalg::GenericOp>(
+        loc, concatTruncType, newGenericOp->getResults(), truncOuts,
+        truncIndexingMaps, truncIteratorTypes, truncateOpBody);
+
+    fusedResult = truncateOp.getResult(0);
+  }
+
+  SmallVector<Value> newOuts;
+
+  auto fusedResultType = cast<RankedTensorType>(fusedResult.getType());
+  SmallVector<int64_t> shape = llvm::to_vector(fusedResultType.getShape());
+
+  SmallVector<OpFoldResult> sizes = newEmpty.getMixedSizes();
+  sizes[0] = rewriter.getIndexAttr(1);
+
+  SmallVector<OpFoldResult> offsets(fusedResultType.getRank(),
+                                    rewriter.getIndexAttr(0));
+  SmallVector<OpFoldResult> strides(fusedResultType.getRank(),
+                                    rewriter.getIndexAttr(1));
+  auto resultOutType = RankedTensorType::get(outType.getShape(),
+                                             fusedResultType.getElementType());
+  for (auto i : llvm::seq<int>(0, rhsVals.size())) {
+    offsets[0] = rewriter.getIndexAttr(i);
+    newOuts.push_back(rewriter.create<tensor::ExtractSliceOp>(
+        loc, resultOutType, fusedResult, offsets, sizes, strides));
+  }
+
+  for (auto [index, op, replacement] :
+       llvm::enumerate(fusionGroup.contractionOps, newOuts)) {
+    Operation *replacedOp = op;
+    if (fusionGroup.truncateOps) {
+      replacedOp = fusionGroup.truncateOps.value()[index];
+    }
+    rewriter.replaceOp(replacedOp, replacement);
+  }
+  return success();
+}
+
+void FuseHorizontalContractionsPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  DominanceInfo dominanceInfo(getOperation());
+
+  SmallVector<HorizontalFusionGroup> horizontalFusionGroups;
+  llvm::SmallDenseSet<linalg::LinalgOp> groupedOperations;
+
+  getOperation()->walk([&](linalg::LinalgOp linalgOp) {
+    if (!linalg::isaContractionOpInterface(linalgOp)) {
+      return;
+    }
+    // Avoid already grouped operations;
+    if (groupedOperations.contains(linalgOp)) {
+      return;
+    }
+
+    std::optional<linalg::GenericOp> truncUser = getTruncUser(linalgOp);
+
+    std::optional<HorizontalFusionGroup> fusionGroup =
+        getHorizontalFusionGroupMembers(linalgOp, truncUser, groupedOperations,
+                                        dominanceInfo);
+
+    if (!fusionGroup) {
+      return;
+    }
+    groupedOperations.insert(fusionGroup->contractionOps.begin(),
+                             fusionGroup->contractionOps.end());
+    horizontalFusionGroups.emplace_back(std::move(fusionGroup.value()));
+  });
+
+  if (horizontalFusionGroups.empty()) {
+    return;
+  }
+
+  IRRewriter rewriter(context);
+  for (auto &fusionGroup : horizontalFusionGroups) {
+    if (failed(fuseGroup(rewriter, fusionGroup))) {
+      return signalPassFailure();
+    }
+  }
+
+  RewritePatternSet patterns(context);
+  tensor::populateDecomposeTensorConcatPatterns(patterns);
```

Comment:
nit: Add
```suggestion
  // Note: Currently these patterns are required due to early lowering of tensor.concat. When we choose
  // to move the lowering of tensor.concat later, these patterns should be dropped.
  tensor::populateDecomposeTensorConcatPatterns(patterns);
```

or alternatively add a pass for decomposing concat to the pipeline.

Reply:
I decided to not do this right now, but it might be better to just use `tensor.insert_slices` directly. Then you can avoid all the expand shape stuff and just use rank-reduced insert slices.